File: typescript/src/daytona/sandbox.ts
import { Arker, ArkerError, type Computer } from "../index.js";
import { FileSystem } from "./files.js";
import { Process } from "./process.js";
import { SandboxState, translateArkerError } from "./types.js";

export interface SandboxConstructorOpts {
  env?: Record<string, string>;
  labels?: Record<string, string>;
  snapshot?: string;
  /** Optional VmInfo-shaped payload from a `arker.get(...)` / list response.
   * If provided, `state` and `snapshot` are populated from it — no extra HTTP. */
  info?: { state?: string; source_golden?: string };
}

function arkerStateToSandboxState(state: string | undefined): SandboxState {
  if (state === "running") return SandboxState.Started;
  if (state === "stopped") return SandboxState.Stopped;
  if (state === "creating" || state === "starting") return SandboxState.Starting;
  if (state === "error") return SandboxState.Error;
  return SandboxState.Unknown;
}

/** daytona.Sandbox drop-in, backed by an Arker `Computer`. Properties are
 * plain assignable fields; reading them doesn't issue HTTP. */
export class Sandbox {
  readonly _arker: Arker;
  readonly _computer: Computer;
  readonly _env: Record<string, string>;
  readonly id: string;
  labels: Record<string, string>;
  snapshot?: string;
  state: SandboxState;
  user = "user";
  public = false;
  readonly target: string;

  readonly process: Process;
  readonly fs: FileSystem;

  constructor(arker: Arker, computer: Computer, opts: SandboxConstructorOpts = {}) {
    this._arker = arker;
    this._computer = computer;
    this._env = { ...(opts.env ?? {}) };
    this.id = computer.id;
    this.labels = { ...(opts.labels ?? {}) };
    this.snapshot = opts.snapshot;
    this.state = SandboxState.Unknown;
    this.target = arker.region ?? "";

    if (opts.info) {
      this.state = arkerStateToSandboxState(opts.info.state);
      if (this.snapshot == null && typeof opts.info.source_golden === "string") {
        this.snapshot = opts.info.source_golden;
      }
    }

    this.process = new Process(this);
    this.fs = new FileSystem(this);
  }

  /** Reads `env` for parity with daytona's assignable-attribute pattern. */
  get env(): Record<string, string> {
    return this._env;
  }

  // Lazy getters retained for back-compat with Phase E test code.
  async getState(): Promise<SandboxState> {
    await this.refreshData();
    return this.state;
  }

  async getSnapshot(): Promise<string | null> {
    if (this.snapshot != null) return this.snapshot;
    try {
      const info = (await this._arker.get(this._computer.id)) as { source_golden?: string };
      this.snapshot = info.source_golden;
      return info.source_golden ?? null;
    } catch {
      return null;
    }
  }

  // ---- Lifecycle ----

  async delete(_timeout: number = 60): Promise<void> {
    try {
      await this._computer.delete();
    } catch (error) {
      throw translateArkerError(error);
    }
    this.state = SandboxState.Destroyed;
  }

  async start(): Promise<void> { /* no-op: Arker VMs run on fork */ }
  async stop(): Promise<void> { /* no-op: Arker has no stopped state */ }
  async archive(): Promise<void> { /* no-op */ }

  // ---- Configuration ----

  setLabels(labels: Record<string, string>): Record<string, string> {
    this.labels = { ...labels };
    return { ...this.labels };
  }

  getUserHomeDir(): string {
    return "/home/user";
  }

  getWorkDir(): string {
    return "/home/user";
  }

  async refreshData(): Promise<void> {
    try {
      const info = (await this._arker.get(this._computer.id)) as { state?: string; source_golden?: string };
      this.state = arkerStateToSandboxState(info.state);
      if (this.snapshot == null && typeof info.source_golden === "string") {
        this.snapshot = info.source_golden;
      }
    } catch (error) {
      if (error instanceof ArkerError) throw translateArkerError(error);
      throw error;
    }
  }
}

File: python/src/arker/daytona/_sandbox.py
"""`daytona.Sandbox` — backed by an Arker `Computer`."""

from __future__ import annotations

import logging
from typing import Any

from ..computer import Arker, ArkerError, Computer
from ._files import FileSystem
from ._process import Process
from ._types import SandboxState

logger = logging.getLogger("arker.daytona")


class Sandbox:
    """daytona.Sandbox drop-in. Many properties (cpu, memory, labels, ...)
    are populated from Arker `VmInfo` where possible; defaults otherwise.

    Pending behaviors live in `__init__.py` (top of module). Notable Phase A
    no-ops: `start`/`stop`/`archive`/`set_*` lifecycle is mostly cosmetic
    because Arker VMs don't have a separate stopped state today.
    """

    def __init__(
        self,
        arker: Arker,
        computer: Computer,
        *,
        env: dict[str, str] | None = None,
        labels: dict[str, str] | None = None,
        snapshot: str | None = None,
    ) -> None:
        self._arker = arker
        self._computer = computer
        self._env: dict[str, str] = dict(env or {})
        self._labels: dict[str, str] = dict(labels or {})
        self._snapshot_id = snapshot
        self.process = Process(self)
        self.fs = FileSystem(self)

    # ---- Properties ----

    @property
    def id(self) -> str:
        return self._computer.id

    @property
    def env(self) -> dict[str, str]:
        return dict(self._env)

    @property
    def labels(self) -> dict[str, str]:
        return dict(self._labels)

    @property
    def state(self) -> SandboxState:
        try:
            arker_state = self._arker.get(self._computer.id).state
        except ArkerError:
            return SandboxState.ERROR
        if arker_state == "running":
            return SandboxState.STARTED
        if arker_state == "stopped":
            return SandboxState.STOPPED
        return SandboxState.ERROR

    @property
    def snapshot(self) -> str | None:
        if self._snapshot_id is not None:
            return self._snapshot_id
        try:
            return self._arker.get(self._computer.id).source_golden
        except ArkerError:
            return None

    @property
    def user(self) -> str:
        return "user"

    @property
    def public(self) -> bool:
        return False

    @property
    def target(self) -> str:
        return self._arker.region or ""

    # ---- Lifecycle ----

    def delete(self, timeout: float | None = 60) -> None:
        try:
            self._computer.delete()
        except ArkerError as error:
            logger.debug("arker.daytona: delete() raised %s", error)

    def start(self, timeout: float | None = 60) -> None:
        # Arker VMs are running on fork; no separate start state. No-op.
        logger.debug("arker.daytona: start() — no-op; Arker VMs are running on fork")

    def stop(self, timeout: float | None = 60, force: bool = False) -> None:
        logger.debug("arker.daytona: stop() — no-op; Arker has no VM stop state")

    def archive(self) -> None:
        logger.debug("arker.daytona: archive() — no-op")

    # ---- Configuration ----

    def set_labels(self, labels: dict[str, str]) -> dict[str, str]:
        # Stored locally; Arker doesn't persist labels server-side.
        self._labels = dict(labels)
        return dict(self._labels)

    # ---- Info helpers ----

    def get_user_home_dir(self) -> str:
        return "/home/user"

    def get_work_dir(self) -> str:
        return "/home/user"

    def refresh_data(self) -> None:
        # Arker has no batched-refresh equivalent; properties read live.
        return None

    # ---- Context manager ----

    def __enter__(self) -> "Sandbox":
        return self

    def __exit__(self, *_exc: Any) -> None:
        try:
            self.delete()
        except Exception:
            pass
